### PR TITLE
fix GradBucket.is_last() logic

### PR DESCRIPTION
**Pull request by nimaelyasi**
Summary: passed number of buckets to GradBucket constructor, to check if index is equal to num_buckets - 1 in the .is_last() function.

Differential Revision: D30455913

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @agolynski @SciPioneer @H-Huang @mrzzd @cbalioglu @gcramer23

### CODE REVIEW

**Review by mrshenli — COMMENTED** [review]
Hey @nimaelyasi, could you please remove the test plan from the PR description, as that's internal-only. We can leave the diff number there, so that reviewers can quickly jump between diff and PR. Thanks!

**Review by mrshenli — COMMENTED** [review]
Hey @nimaelyasi, thanks for submitting this fix. After addressing internal comments, please also address internal and OSS lint errors. Thanks!

**Review by wayi1 — APPROVED** [review]

**Review by mrshenli — COMMENTED** [review]
Thanks @nimaelyasi for working on this. I made some minor suggestions.

File: test/distributed/algorithms/ddp_comm_hooks/test_ddp_hooks.py [hunk 1]
```diff
@@ -177,6 +191,61 @@ def test_ddp_comm_hook_quantize_per_channel_hook(self):
...
+            )
+
+            # Register DDP Communication Hook if defined
+            if hook is not None:
+                gpu_model.register_comm_hook(state=process_group, hook=hook)
+
+            return my_run_and_get_grads(gpu_model)
+
+        # Register hook case, get the hook grads.
+        hook_grads = my_get_grads(process_group, reset_last_bucket_grad_hook)
```
[the remark this answers is not in the document]
Comment (nimaelyasi) [reply test/distributed/algorithms/ddp_comm_hooks/test_ddp_hooks.py]: Hi @mrshenli, thanks for your suggestion. I made the changes as you suggested. Please let me know if any further changes needed. Thanks!

**Review by mrshenli — APPROVED** [review]
LGTM! Thanks!

**Pull request closed without merging**